Add parser for dispatch language

andywingo opened the pull request:
See #229.

* src/pf/parse.lua: Add options dict, and ability to parse arithmetic
  expressions directly.

* src/pf/dispatch.lua: New file.

* src/Makefile: Add new file.

kbara commented on src/pf/dispatch.lua:
```diff
@@ -0,0 +1,164 @@
+local set, assert_equals = utils.set, utils.assert_equals
+local parse_pflang = require('pf.parse').parse
+
+local function split(str, pat)
```
I wonder whether split belongs in utils.lua.

kbara commented on src/pf/dispatch.lua:
```diff
@@ -0,0 +1,164 @@
+local function scanner(str)
+   local pos = 1
+   local function error_str(message, ...)
+      local location_error_message = "Parse error: In expression \"%s\""
```
Please make this more explicit, ie by saying "Pfmatch parse error". We've already had an issue filed when it wasn't clear what layer an error message was coming from.

kbara commented:
The implementation looks sane. If you address the parse nit and add a big comment about how this is an experimental feature that no one else should rely on yet, LGTM to merge for prototyping if the language itself is a good idea.

andywingo commented:
Updated patch renames dispatch.lua to "switch.lua", adds support for comments, and gives better errors.  I had a hard time making things coherently named so I just changed most things to "switch".  We know that a witch is inside!

kbara commented:
Sorry, I totally have to bikeshed calling it switch.lua. That term is way too overloaded in the context of this being something we'll use with Snabbswitch. I'd be ok with dispatch.lua, match.lua, pfiff.lua, maybe witch.lua, or lots of other options, but I think switch.lua is probably a bit too misleading.

andywingo commented:
Dunno, the name is designed to be overloaded with "switch" in programming languages.  Which would you rather see in a program:

```
switch (pass, drop) { udp=>pass, tcp=>drop }
dispatch (pass, drop) { udp=>pass, tcp=>drop }
match (pass, drop) { udp=>pass, tcp=>drop }
pfiff (pass, drop) { udp=>pass, tcp=>drop }
witch (pass, drop) { udp=>pass, tcp=>drop }
```

In this context I think switch has a lot to recommend itself!

kbara commented:
Of those options, I'd prefer 'dispatch', I think. 'match' is more suggestive of potentially matching multiple times, while 'switch' makes me think of C's fallthrough semantics. 'pfiff' and 'witch' are both cutsie to the point of not really conveying semantics well; they'd be better as language names than as a keyword, though they share the drawback there too.

andywingo commented:
Updated branch renames to "match" throughout, makes => after the 'match()' unnecessary.

andywingo commented:
Updated branch adds "otherwise", implements expander.  Will see if I can do the optimizer and codegen this morning too.

andywingo commented:
Having to list handlers twice is a bit of a drag.  Instead let's have the matcher take a table as an argument, and then dispatch to calls to members of that table.  Like:

```lua
local matcher = pf.match.compile([[match {
   ip => pass
}]])
local handlers = {
   pass = function(data, len) return true end
}
local match = matcher(handlers)
...
if match(pkt.data, pkt.len) then link.transmit(out, pkt) end
```

Dunno.  It's more dynamic binding (what before was a call to a known closure var is now a table access, probably a shape check, ...) and you get less static checking (though we could rectify that to an extent)... but it's fewer live variables and more pleasant to work with.

Alternately... we could pass the handler table to the matcher every time.  So like this:

```lua
local matcher = pf.match.compile([[match {
   ip => pass
}]])
local handlers = {
   pass = function(data, len) return true end
}
...
if match(handlers, pkt.data, pkt.len) then link.transmit(out, pkt) end
```

andywingo commented:
See https://github.com/Igalia/snabbswitch/pull/10/files for a sketch of what this new thing would look like in practice

kbara commented:
Agreed about listing handlers twice; I was hoping that would disappear. I prefer explicitly passing handlers to match, because it's useful to be able to match against more than one set of handlers in the same source file.

andywingo commented:
FWIW if you match against more than one set of handlers in a source file that would kill perf, as all those handlers.foo() calls become non-monomorphic.  Best practice is always going to be "call one matcher with one handler obj".

Incidentally, the handlers dict can be the app itself:

```lua
local app = {
   matcher = pf.match.compile([[match {
      ip => pass
   }]]),
   pass = function(data, len) return true end
}

...
function App:push()
   ...
   if self:match(pkt.data, pkt.len) then link.transmit(out, pkt) end
end
```

There's potential for confusion here between handlers and non-handlers but for short apps where you can hold the whole program in your head, this might be a good option.

kbara commented:
Hmm. Fair points. Given that, I prefer the first option, because that makes this particular trap harder to fall into.

andywingo commented:
Going ahead with handler table as an argument for now.  It's easy enough to bind the handler table statically:

```lua
function make_matcher(prog_str)
   return function(handlers)
      local matcher = pf.match.compile(prog_str)
      return function(data, len)
         return matcher(handlers, data, len)
      end
   end
end
```

andywingo commented:
Here's the result of compiling this expression:

```
match { tcp port 80 => pass }
```

```lua
local lshift = require("bit").lshift
local band = require("bit").band
local cast = require("ffi").cast
return function(self,P,length)
   if length < 34 then return false end
   local var1 = cast("uint16_t*", P+12)[0]
   if var1 == 8 then
      if P[23] ~= 6 then return false end
      if band(cast("uint16_t*", P+20)[0],65311) ~= 0 then return false end
      local var7 = lshift(band(P[14],15),2)
      local var8 = (var7 + 16)
      if var8 > length then return false end
      if cast("uint16_t*", P+(var7 + 14))[0] == 20480 then goto L6 end
      if (var7 + 18) > length then return false end
      if cast("uint16_t*", P+var8)[0] == 20480 then goto L6 end
      goto L7
   else
      if var1 ~= 56710 then return false end
      if length < 54 then return false end
      local var24 = P[20]
      if var24 == 6 then goto L24 end
      do
         if var24 ~= 44 then return false end
         if length < 55 then return false end
         if P[54] == 6 then goto L24 end
         return false
      end
::L24::
      if length < 56 then return false end
      if cast("uint16_t*", P+54)[0] == 20480 then goto L6 end
      if length < 58 then return false end
      if cast("uint16_t*", P+56)[0] == 20480 then goto L6 end
      goto L7
   end
::L6::
   do
      return self.pass(P, len)
   end
::L7::
   return false
end
```

andywingo commented:
The control flow is pretty gross; we need to improve it.  Like no one should be jumping to L7 which only does a return false.  Anyway, I think this is probably something that pflang already has so the matcher doesn't add anything.

andywingo commented:
And here's the output for the match expression in the basicnat-dispatch branch:

```lua
match = require('pf.match')
print(match.compile([[match {
   not ip => forward
   -- Drop fragmented packets.
   ip[6:2] & 0x3fff != 0 => drop
   ip src $external_ip => {
      tcp => incoming_tcp(&ip[0], &tcp[0])
      udp => incoming_udp(&ip[0], &udp[0])
      otherwise => incoming_other(&ip[0])
   }
   ip dst $internal_ip => {
      tcp => outgoing_tcp(&ip[0], &tcp[0])
      udp => outgoing_udp(&ip[0], &udp[0])
      otherwise => outgoing_other(&ip[0])
   }
   otherwise => drop
}]], {subst={external_ip='1.2.3.4', internal_ip='5.6.7.8'}, source=true}))
```

prints

```lua
local lshift = require("bit").lshift
local band = require("bit").band
local cast = require("ffi").cast
return function(self,P,length)
   if length < 14 then return false end
   if cast("uint16_t*", P+12)[0] == 8 then
      if length < 34 then return false end
      local var2 = cast("uint16_t*", P+20)[0]
      if band(var2,65343) ~= 0 then
         return self.drop(P, len)
      end
      if cast("uint32_t*", P+26)[0] == 67305985 then
         local var5 = P[23]
         if var5 ~= 6 then goto L15 end
         do
            if length < 54 then goto L15 end
            if band(var2,65311) ~= 0 then goto L15 end
            if (lshift(band(P[14],15),2) + 15) > length then goto L15 end
            return self.incoming_tcp(P, len, 14, (lshift(band(P[14],15),2) + 14))
         end
::L15::
         if var5 ~= 17 then goto L23 end
         do
            if length < 42 then goto L23 end
            if band(var2,65311) ~= 0 then goto L23 end
            if (lshift(band(P[14],15),2) + 15) > length then goto L23 end
            return self.incoming_udp(P, len, 14, (lshift(band(P[14],15),2) + 14))
         end
::L23::
         return self.incoming_other(P, len, 14)
      else
         if cast("uint32_t*", P+30)[0] == 134678021 then
            local var28 = P[23]
            if var28 ~= 6 then goto L33 end
            do
               if length < 54 then goto L33 end
               if band(var2,65311) ~= 0 then goto L33 end
               if (lshift(band(P[14],15),2) + 15) > length then goto L33 end
               return self.outgoing_tcp(P, len, 14, (lshift(band(P[14],15),2) + 14))
            end
::L33::
            if var28 ~= 17 then goto L41 end
            do
               if length < 42 then goto L41 end
               if band(var2,65311) ~= 0 then goto L41 end
               if (lshift(band(P[14],15),2) + 15) > length then goto L41 end
               return self.outgoing_udp(P, len, 14, (lshift(band(P[14],15),2) + 14))
            end
::L41::
            return self.outgoing_other(P, len, 14)
         else
            return self.drop(P, len)
         end
      end
   else
      return self.forward(P, len)
   end
end
```

andywingo commented:
I guess the tcp[0] stuff restricts the packet to be ipv4 only, but that's ok for that particular example.

andywingo commented:
Okeydoke, I think we're ready to go.  WDYT @kbara?

kbara commented on doc/pfmatch.md:
```diff
@@ -0,0 +1,234 @@
+
+Here's a more complicated example that passes all non-IP traffic, drops
+all IP traffic that is not going to or coming from certain IP addresses,
+and calls a handler on the rest of the traffic.
```
Nit: it also drops fragmented packets.

andywingo replied:
I'll fix to remove that

kbara commented on doc/pfmatch.md:
````diff
@@ -0,0 +1,234 @@
+   ip dst 5.6.7.8 => outgoing_ip(&ip[0])
+   otherwise => drop
+}
+```
````
It's probably worth documenting what &ip[0] is in this context: how does it differ from ip[0]?
Also, what are the arguments to drop vs incoming_ip? Naively, one looks like a function call and one doesn't, and it's worth making it clear that incoming_ip(...) isn't returning a higher order function to be dispatched to, for instance. The implicit/explicit mix here and how it compares to normal Lua is a bit disconcerting without a bit more explanation.

andywingo replied:
Dunno, you don't find the two sentences preceding the example as sufficiently explanatory?

kbara commented on doc/pfmatch.md:
```diff
@@ -0,0 +1,234 @@
+match {
+   not ip => forward
+   -- Drop fragmented packets.
+   ip[6:2] & 0x3fff != 0 => drop
```
It's probably worth making it explicit that at most one clause is executed, so fragmented packets are dropped regardless of their src/dst IPs.

kbara commented on doc/pfmatch.md:
```diff
@@ -0,0 +1,234 @@
+itself, and you're missing out on opportunities for optimization.  For
+example, the if the packet fails the `not_ip` check, then we don't need
+to check if it's an IP packet in the `fragmented` check.  Compiling a
+pfmatch program takes advantage of pflua's optimizer to produce optimal
```
Rather good code, yes. Optimal code? Possibly slightly overselling.

kbara commented:
LGTM after doc nits are addressed, with the caveat that I've only read the code, not used it. :)

andywingo commented:
Updated patch fixes handling of "fail" conditions in pflang, tweaks a couple of optimizations, and then updates docs.  There's still something not quite right about the length hoisting though.  I'm going to merge this then look at length hoisting to see what's going wrong.
